bindings/pyroot/pythonizations/python/ROOT/pythonization/_roofit/_utils.py

```diff
@@ -74,6 +81,82 @@ def _string_to_root_attribute(value, lookup_map):
         return value
 
 
+def _dict_to_std_map(arg_dict, allowed_val_dict):
+    """Helper function to convert python dict to std::map."""
+    import ROOT
+
+    def all_of_class(d, type, check_key):
+        return all([isinstance(key if check_key else value, type) for key, value in d.items()])
+
+    def get_python_class(cpp_type_name):
+
+        cpp_type_name = cpp_type_name.replace("*", "")
+
+        if cpp_type_name in ["std::string", "string"]:
+            return str
+        if cpp_type_name == "int":
+            return int
+
+        # otherwise try to get class from the ROOT namespace
+        return getattr(ROOT, cpp_type_name)
+
+    def get_template_args(import_dict):
+
+        key_type = None
+        value_type = None
+
+        for typename in allowed_val_dict.keys():
+            if all_of_class(import_dict, get_python_class(typename), True):
+                key_type = typename
+
+        for typename in allowed_val_dict.values():
+            if isinstance(typename, list):
+                for type_val in typename:
+                    if all_of_class(import_dict, get_python_class(type_val), False):
+                        value_type = type_val
+
+            if all_of_class(import_dict, get_python_class(typename), False):
+                value_type = typename
+
+        def get_python_typenames(typenames):
+            return [get_python_class(t).__name__ for t in typenames.values()]
+
+        def prettyprint_str_list(l):
+            if len(l) == 1:
+                return l[0]
+            if len(l) == 2:
+                return l[0] + " or " + l[1]
+            return ", ".join(l[:-1]) + ", or " + l[-1]
+
+        if key_type is None:
+            raise TypeError(
+                "All dictionary keys must be of the same type, which can be either "
+                + prettyprint_str_list(get_python_typenames(allowed_val_dict))
+                + "."
+            )
+
+        if value_type is None:
+            raise TypeError(
+                "All dictionary values must be of the same type, which can be either "
+                + prettyprint_str_list(get_python_typenames(allowed_val_dict))
+                + "."
+            )
+
+        return key_type + "," + value_type
```

Comment:
This is still not how @etejedor suggested it and how we discussed it.

The idea was that for each allowed key type, you have one ore more allowed value types in a list. This means you should have a nested loop, with the outer loop for the key type and the inner for the value type. In your current implementation, you are just looping over all possible value types and ignore the relation to the key type.

To speed things up a bit, here is how I would suggest to implement it, also with a more meaningful error message if no match for the value type is found:

```Python
    def prettyprint_str_list(l):
        if len(l) == 1:
            return l[0]
        if len(l) == 2:
            return l[0] + " or " + l[1]
        return ", ".join(l[:-1]) + ", or " + l[-1]


    def get_template_args(import_dict):

        key_type = None
        value_type = None

        def get_python_typenames(typenames):
            return [get_python_class(t).__name__ for t in typenames]

        for key_typename in allowed_val_dict.keys():
            if all_of_class(import_dict, get_python_class(key_typename), True):
                key_type = key_typename

                if type(allowed_val_dict[key_typename]) == str:
                    allowed_val_dict[key_typename] = [allowed_val_dict[key_typename]]

                for val_typename in allowed_val_dict[key_typename]:
                    if all_of_class(import_dict, get_python_class(val_typename), False):
                        value_type = val_typename

                if value_type is None:
                    raise TypeError(
                        "All dictionary values must be of the same type, which can be either "
                        + prettyprint_str_list(get_python_typenames(allowed_val_dict[key_typename]))
                        + ", given the key type " + get_python_class(key_type).__name__ + "."
                    )

        if key_type is None:
            raise TypeError(
                "All dictionary keys must be of the same type, which can be either "
                + prettyprint_str_list(get_python_typenames(allowed_val_dict))
                + "."
            )

        return key_type + "," + value_type
```